### PR TITLE
[Broker] Support create subscription by specifying the earliest or latest position.

### LINKED ISSUE #9441: Creating subscription on a Topic via REST API sets the cursor to earliest position
**Describe the bug**
When creating a subscription using REST API on an existing topic having a good flow of messages, the cursor is set to the earliest position and if the same is done via pulsar-admin it sets the cursor correctly.

**To Reproduce**
1. Create a topic having messages flowing in continuously and create a consumer for it.
2. Let some messages get accumulated.
3. Create a new subscription via REST API using this REST API
    ```
    curl -X PUT -L http://<pulsar_url>:8080/admin/v2/persistent/<tenant>/<namespace>/<topic>/subscription/<subscription_name>_rest_api -H 'Content-Type: application/json' -d '{ "messageId": "latest"}'
    ```
4. Also create a subscription using Admin API
    ```
    pulsar-admin topics create-subscription -m latest -s <subscription_name>_admin_api persistent://<tenant>/<namespace>/<topic>
    ```

**Expected behavior**
The subscription created via REST API should also have the cursor set on the latest.

**Screenshots**
Both subscription_adminApi and subscription_restApi were created at the same time. See the difference in backlog.
Screenshot 2021-02-03 at 10 08 33 AM

### PR DESCRIPTION
**Pull request by RobertIndie**
Fixes #9441

### Motivation

Currently, if we need to specify the messageID when creating a subscription through the REST API, we cannot use `latest` and `earliest` to specify the position. Instead, we can only do it like this to create a subscription from the latest position:
```
PUT /admin/v2/persistent/public/default/test/subscription/test HTTP/1.1
Content-Type: application/json

{"ledgerId":9223372036854775807,"entryId":9223372036854775807,"partitionIndex":-1}
```

This PR adds support for creating subscription by specifying the `latest` or `earliest`  position. So we can send the request like this:
```
PUT /admin/v2/persistent/public/default/test/subscription/test HTTP/1.1
Content-Type: application/json

"latest"
```

### Modifications

* Add a private constructor for MessageIdImpl so that we can convert the string parameter like `latest` or `earliest` to the MessageIdImpl.

### Verifying this change

This change is already covered by existing tests, such as *createSubscriptionBySpecifyingStringPosition*.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (yes)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

  
- [x] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java [hunk 1]
```diff
@@ -45,6 +45,18 @@ private MessageIdImpl() {
         this(-1, -1, -1);
     }
 
+    // Private constructor used only for json deserialization
+    private MessageIdImpl(String position) {
```
Comment (nlu90) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:49]: should we check any invalid value other than "earliest" and "latest"?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 2]
```diff
@@ -1318,9 +1318,9 @@ public void createSubscription(
             @ApiParam(value = "Is authentication required to perform this operation")
             @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
             @ApiParam(name = "messageId", value = "messageId where to create the subscription. "
-                    + "It can be 'latest', 'earliest' or (ledgerId:entryId)",
+                    + "It can be 'latest', 'earliest' or (ledgerId:entryId:partitionIndex)",
                     defaultValue = "latest",
-                    allowableValues = "latest, earliest, ledgerId:entryId"
+                    allowableValues = "latest, earliest, (ledgerId:entryId:partitionIndex)"
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:1323]: We should use the `ResetCursorData` here, for couple reasons:

1. In pulsar, the default behavior is to create the subscription on the latest position, so if users did not provide the messageId, it means creating a subscription by using the latest position.
2. `isExcluded` and `batchIndex` should be also considered when creating a subscription.
3. Follow the same API def with PersistentTopics.resetCursorOnPosition(), for passing the position information, we should provide a consistent way.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java (hunk 1 above)
Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java:49]: Yes. I think we can throw an exception in the constructor to tell the user the position is invalid.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 2 above)
Comment (RobertIndie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:1323]: I agree. `ResetCursorData` is more suitable than `MessageIdImpl`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 3]
```diff
@@ -1322,7 +1322,7 @@ public void createSubscription(
                     defaultValue = "latest",
                     allowableValues = "latest, earliest, ledgerId:entryId"
             )
-                    MessageIdImpl messageId,
+                    ResetCursorData resetCursorData,
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:1325]: Oh, Sorry @RobertIndie my mistake, it looks like we already have `messageId` there, changing to resetCursorData will introduce compatibility issues.

**Review by nlu90 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 3 above)
Comment (RobertIndie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:1325]: This is the REST API. The Json format of the `ResetCursorData` and `MessageIdImpl` used here is the same. It looks like there are no compatibility issues here.

**Comment by Anonymitaet** [comment]
Confirmed w/ @RobertIndie, doc is the annotations generated from code files.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ResetCursorData.java [hunk 4]
```diff
@@ -43,6 +43,21 @@ public ResetCursorData(long ledgerId, long entryId, boolean isExcluded) {
...
+    // Private constructor used only for json deserialization
+    private ResetCursorData(String position) {
+        if ("latest".equals(position)) {
+            this.ledgerId = Long.MAX_VALUE;
+            this.entryId = Long.MAX_VALUE;
+        } else if ("earliest".equals(position)) {
+            this.ledgerId = -1;
+            this.entryId = -1;
+        } else {
+            throw new IllegalArgumentException(
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ResetCursorData.java:55]: It looks like we are not supporting the same encoding of MessageIdImpl and you are only supporting the two constant values.

Am I missing something?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ResetCursorData.java (hunk 4 above)
Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ResetCursorData.java:55]: Just like the MessageIdImpl, the encoding format like `{"ledgerId":9223372036854775807,"entryId":9223372036854775807,"partitionIndex":-1}
` is not implemented through this constructor, but directly through deserialization of the json object. This constructor is only for the case of constant values.

Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ResetCursorData.java:55]: @eolivelli PTAL.

**Review by eolivelli — APPROVED** [review]
+1

**Pull request merged**